packages/web-components/fast-html/RENDERING.md

```diff
@@ -0,0 +1,311 @@
+    }
+
+    hasStateAndShadowOptions() {
+        if (this.shadowOptions && this.initialState) {
+            this.deferHydration = false;
+        }
+    }
+
+    loadItemFromInitialState(): void {
+        const initialState = this.initialStateService.init(["text"]);
+
+        this.text = initialState.text;
+
+        this.initialState = true;
+
+        this.hasStateAndShadowOptions();
+    }
+}
+
+MyComponent.define({
+    name: "my-component",
+    shadowOptions: null
+});
+```
+
+Notice that there is a `defer-hydration` attribute that gets removed once initial state has been applied and `shadowOptions` has also been defined. This allows us to ensure that the template has been found and parsed to the component and the initial state has been applied before we connect the component to the rendered HTML.
+
+Depending on your application structure it might be more maintainable to create an abstract class that extends `FASTElement` to re-use the service logic and the removal of `defer-hydration`.
+
+## Hydration Comments and Datasets
+
+When hydrating the HTML FAST uses the `HydratableElementController` which can be included in your bundle like so:
+
+```typescript
+import "@microsoft/fast-element/install-element-hydration.js";
+```
+
+The `HydratableElementController` will take over from the `ElementController` to use hydration "markers" such as comments and dataset attributes to rationalize bindings with existing HTML.
+
+### Bindings
+
+#### Syntax
+
+A content binding syntax can be broken into parts:
+- `fe-b` - declares this as a FASTElement Binding.
+- `start|end` - indicates the start or end of the binding.
+- `number` - the number in order in the template of bindings, these increment for every binding.
+- `UUID` - a unique string identifying the binding. The examples shown here have been generated from a 10 character long string used in the [FAST SSR package](https://github.com/microsoft/fast/blob/main/packages/web-components/fast-ssr/src/template-parser/id-generator.ts).
+
+Separator:
+- `$$` - a separator string between the parts.
+
+An attribute binding is tracked using a dataset attribute with the name `data-fe-b` followed by the binding number.
+
+#### Examples
+
+**Simple content example**
+
+Content bindings such as:
+```html
+<h1>{{title}}</h1>
+```
+
+When combined with state such as:
+```json
+{
+    "title": "Hello world"
+}
+```
+
+Should result in:
+```html
+<h1><!--fe-b$$start$$0$$ZJEYduCZlM$$fe-b-->Hello world<!--fe-b$$end$$0$$ZJEYduCZlM$$fe-b--></h1>
+```
+
+**Simple attribute example**
+
+Attribute binding such as:
+```html
+<h1 title="{{title}}"></h1>
+```
+
+Should result in:
+
+```html
+<h1 data-fe-b="0" title="Hello world"></h1>
+```
+
+If multiple attribute bindings exist on the same element, these will be separated by a space.
+
+Example:
+```html
+<h1 title="{{title}}" subtitle="{{subtitle}}"></h1>
+```
+
+Expected result:
+```html
+<h1 data-fe-b="0 1" title="Hello" subtitle="world"></h1>
+```
```

Comment:
Question - should we use a custom elements here since custom attributes don't exist yet for standard HTML elements? I'm also curious if `title` could confuse given the actual `title` attribute. Thoughts?

Reply:
Good point on the `title` attribute, I'll change that to something not part of the native HTMLElement spec. I could wrap these with the custom element tags to further indicate what needs to be generated, I was doing a 1 to 1 conversion with FAST HTML internal declarative HTML syntax and the expected generated output.